core/analysis/collation_token_stream.cpp

```diff
@@ -0,0 +1,354 @@
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Andrey Abramov
+////////////////////////////////////////////////////////////////////////////////
+
+#include "collation_token_stream.hpp"
+
+#include <unicode/locid.h>
+#include <unicode/coll.h>
+
+#include "velocypack/Slice.h"
+#include "velocypack/Builder.h"
+#include "velocypack/Parser.h"
+#include "velocypack/velocypack-aliases.h"
+
+#include "utils/locale_utils.hpp"
+#include "utils/vpack_utils.hpp"
+
+namespace iresearch {
+namespace analysis {
+
+struct collation_token_stream::state_t {
+  icu::Locale icu_locale;
+  const options_t options;
+  std::unique_ptr<icu::Collator> collator;
+  std::string utf8_buf;
+  byte_type term_buf[32768];
+
+  state_t(const options_t& opts)
+    : icu_locale("C"), options(opts) {
+    // NOTE: use of the default constructor for Locale() or
+    //       use of Locale::createFromName(nullptr)
+    //       causes a memory leak with Boost 1.58, as detected by valgrind
+    icu_locale.setToBogus(); // set to uninitialized
+  }
+};
+
+} // analysis
+} // iresearch
+
+namespace {
+
+using namespace irs;
+
+bool make_locale_from_name(const string_ref& name,
+                           std::locale& locale) {
+  try {
+    // true == convert to unicode, required for ICU
+    locale = locale_utils::locale(name, string_ref::NIL, true);
+
+    // check if ICU supports locale
+    auto icu_locale =
+        icu::Locale(std::string(locale_utils::language(locale)).c_str(),
+                    std::string(locale_utils::country(locale)).c_str());
+    return !icu_locale.isBogus();
+  } catch (...) {
+    IR_FRMT_ERROR(
+      "Caught error while constructing locale from "
+      "name: %s",
+      name.c_str());
+  }
+  return false;
+}
+
+constexpr VPackStringRef LOCALE_PARAM_NAME{"locale"};
+
+bool parse_vpack_options(
+    const VPackSlice slice,
+    analysis::collation_token_stream::options_t& options) {
+
+  if (!slice.isObject() && !slice.isString()) {
+    IR_FRMT_ERROR(
+      "Slice for delimited_token_stream is not an object or string");
+    return false;
+  }
+
+  try {
+    switch (slice.type()) {
+      case VPackValueType::String:
+        return make_locale_from_name(get_string<string_ref>(slice), options.locale);  // required
+      case VPackValueType::Object:
+      {
+        auto param_name_slice = slice.get(LOCALE_PARAM_NAME);
+        if (!param_name_slice.isNone() && param_name_slice.isString()) {
```

Comment:
```suggestion
        if (param_name_slice.isString()) {
```
If slice isString it is for sure not isNone, so check is redudant